sdk/pre-authorized-debit-v1/src/write/implementation/TransactionFactory.ts

```diff
@@ -0,0 +1,502 @@
+        params.wrapNativeMintParams;
+
+      if (
+        lamportsSourceAccount &&
+        !lamportsSourceAccount.equals(tokenAccount.owner)
+      ) {
+        additionalExpectedSigners.push({
+          publicKey: lamportsSourceAccount,
+          reason: `${lamportsSourceAccount.toBase58()} needs to sign to debit lamports from it`,
+        });
+      }
+
+      setupInstructions.push(
+        SystemProgram.transfer({
+          fromPubkey: lamportsSourceAccount ?? tokenAccount.owner,
+          toPubkey: params.tokenAccount,
+          lamports: wrapLamportsAmount,
+        }),
+      );
+
+      setupInstructions.push(
+        createSyncNativeInstruction(params.tokenAccount, tokenProgramId),
+      );
+    }
+
+    return this.wrapIxsInTx(
+      setupInstructions,
+      [initRecurringPreAuthorizationIx.instruction],
+      undefined,
+      [
+        ...initRecurringPreAuthorizationIx.expectedSigners,
+        ...additionalExpectedSigners,
+      ],
+      initRecurringPreAuthorizationIx.meta,
+    );
+  }
+
+  public async buildDebitTx(
+    params: DebitParams & UnwrapNativeMintAdditionalParams,
+  ): Promise<TransactionWithMetadata<DebitResult>> {
+    const debitIx = await this.ixFactory.buildDebitIx(params);
+
+    const cleanupInstructions: TransactionInstruction[] = [];
+    const additionalExpectedSigners: ExpectedSigner[] = [];
+
+    const preAuthorization = await this.fetchPreAuthorizationOrThrow(
+      params.preAuthorization,
+    );
+
+    const tokenProgramId =
+      await this.readClient.fetchTokenProgramIdForTokenAccount(
+        params.destinationTokenAccount,
+      );
+
+    // NOTE: We don't create this token account, caller has to do it.
+    const destinationTokenAccount = await getAccount(
+      this.connection,
+      params.destinationTokenAccount,
+      undefined,
+      tokenProgramId,
+    );
+
+    if (destinationTokenAccount.isNative && params.unwrapNativeMintParams) {
+      if (
+        !destinationTokenAccount.owner.equals(
+          preAuthorization.account.debitAuthority,
+        )
+      ) {
+        additionalExpectedSigners.push({
+          publicKey: destinationTokenAccount.owner,
+          reason: `The owner of token account ${destinationTokenAccount.address.toBase58()} (owner: ${destinationTokenAccount.owner.toBase58()}) has to sign to close it`,
+        });
+      }
+
+      cleanupInstructions.push(
+        createCloseAccountInstruction(
+          params.destinationTokenAccount,
+          params.unwrapNativeMintParams.lamportsDestinationAccount ??
+            destinationTokenAccount.owner,
+          destinationTokenAccount.owner,
+          undefined,
+          tokenProgramId,
+        ),
+      );
+    }
+
+    return this.wrapIxsInTx(
+      undefined,
+      [debitIx.instruction],
+      cleanupInstructions,
+      debitIx.expectedSigners,
+      debitIx.meta,
+    );
+  }
+
+  private async fetchPreAuthorizationOrThrow(pubkey: PublicKey) {
```

Comment:
Would be more useful to have this as a public method in the readClient and take a custom error as an optional param.